### ISSUE TITLE
X509_check_private_key() crash with EC key from engine

### ISSUE THREAD

**Issue by dwmw2** [openssl/openssl#1532]
(cf. https://github.com/OpenSC/libp11/issues/98)

When the PKCS#11 engine has only private key and no public key, I get a crash in `EC_POINT_cmp()`:

```
Starting program: /home/dwmw2/git/openconnect/ossl/.libs/openconnect -c pkcs11:token=foo1\;pin-value=1234\;object=ecnopub localhost:4443 --servercert sha1:a82547f68f44d6351bef6cacd1d7b96e84f9dfa3
...
Program received signal SIGSEGV, Segmentation fault.
0x00007ffff6c30a58 in EC_POINT_cmp (group=0x6d7680, a=0x6d6660, b=0x0, ctx=0x0)
    at ec_lib.c:1001
1001        if ((group->meth != a->meth) || (a->meth != b->meth)) {
Missing separate debuginfos, use: dnf debuginfo-install krb5-libs-1.14.1-7.fc23.x86_64 libp11-0.3.0-1.fc23.x86_64 libtasn1-4.8-2.fc23.x86_64 lz4-r131-1.fc23.x86_64 opensc-0.15.0-4.fc23.x86_64 pcsc-lite-libs-1.8.15-1.fc23.x86_64 softhsm-2.1.0-1.fc23.x86_64 stoken-libs-0.90-2.fc23.x86_64
(gdb) bt
#0  0x00007ffff6c30a58 in EC_POINT_cmp (group=0x6d7680, a=0x6d6660, b=0x0, 
    ctx=0x0) at ec_lib.c:1001
#1  0x00007ffff6c3aacc in eckey_pub_cmp (a=<optimized out>, b=0x6d44c0)
    at ec_ameth.c:224
#2  0x00007ffff6caee51 in X509_check_private_key (x=<optimized out>, 
    k=k@entry=0x6d44c0) at x509_cmp.c:333
#3  0x00007ffff6ffda9c in ssl_set_pkey (c=0x649bc0, pkey=0x6d44c0)
    at ssl_rsa.c:223
#4  0x00007ffff7bcdf99 in load_pkcs11_key (vpninfo=0x644a90)
    at ../openssl-pkcs11.c:584
#5  0x00007ffff7bca00d in load_certificate (vpninfo=0x644a90)
    at ../openssl.c:889
#6  0x00007ffff7bcb65c in openconnect_open_https (vpninfo=0x644a90)
    at ../openssl.c:1633
#7  0x00007ffff7ba21a0 in do_https_request (vpninfo=0x644a90, 
    method=0x7ffff7bd2bdf "POST", 
    request_body_type=0x7ffff7bd2e60 "application/x-www-form-urlencoded", 
    request_body=0x645660, form_buf=0x7fffffffd960, fetch_redirect=0)
    at ../http.c:875
#8  0x00007ffff7bbba70 in cstp_obtain_cookie (vpninfo=0x644a90)
    at ../auth.c:1285
#9  0x00007ffff7ba5b06 in openconnect_obtain_cookie (vpninfo=0x644a90)
    at ../library.c:182
```

**Comment by dengert** [comment]
https://www.openssl.org/docs/manmaster/crypto/EVP_PKEY_cmp.html
says:
"Since OpenSSL private keys contain public key components too the function EVP_PKEY_cmp() can also be used to determine if a private key matches a public key."

But that is not always possible.

With PKCS#11 the private key does not have an attribute for the EC_POINT. There in no requirement that PKCS#11 has access to the match public key. When using a hardware token, that has an EC key it does not have to expose it. But the token can still do ECDSA sign.

The OpenSC libp11 can create a EVP_KEY and provide a ECDSA_METHOD (ECKEY_METHOD (in 1.1)  that does have the EC_POINT.  


 

In eckey_pub_cmp:223  
  *pb = EC_KEY_get0_public_key(b->pkey.ec);
returns NULL, and then calls 
  r = EC_POINT_cmp(group, pa, pb, NULL);
that then segfaults. 

Possible solutions:
EC_KEY_get0_public_key(b->pkey.ec); returns NULL assume the keys do match and even copy the EC_POINT  to the private key.

**Comment by dwmw2** [comment]
Why assume? We can perform a signature using the privkey and validate it using the cert.

**Comment by dengert** [comment]
You mean add another signature operation to be done on a possibly slow smart card?

Who is "we" in this? There are multiple pieces of software involved that each have a piece of the information and expectations of what they need.

OpenSSL expects it private  EVP_PKEY  to contain the ec_point. 

The token, that does not have or does not expose the ec_point for a private key. 

PKCS#11 does not return the ec_point as an attribute for private key, and specifically says to get the ec_pont from the public key.

libp11 that gets all its information from the token via PKCS#11 API. 

You, the user, said you did not want to put the pubkey on the token. 
If you put the pubkey on the token, the problem goes away. 

The openconnect could try and add the ec_point to the EVP_KEY by parsing the cert you are reading from a file anyway. But that only fixes the issue for openconnect. 

libp11 could fail to load the private key if it can not also find the ec_point via the pubkey. 
If we add this you will have to  put the pubkey on the token anyway. 

In any case OpenSSL should not segfault. But what should OpenSSL do if the the key does not have the ec_point? Why is it comparing the keys to see if they match? Sounds like to me they added extra checks but if they did not check and just went ahead things should work. (Should be verified that that is the case.) But they may stick by their statement that  "OpenSSL private keys contain public key components" and fail the compare. 

If they take that stance, you will have to patch openconnect or add the pubkey to the token. 

I have not run into this situation. The main card I have interest in is the PIV card. The only way to find out what keys are on the card, the type of the key, the parameters, the size and the ec_point is to read the certificate object for each key. The OpenSC driver reads the certificate to get the information and then presents via PKCS#11 the certificate, public key and private key. libp11 then reads the public key and private key and get the ec_point from the public key to place into the EVP_PKEY used by OpenSSL.     
 
 So why not just put the public key on the token?

**Comment by dwmw2** [comment]
_"Why not just put the public key on the token?"_ No reason... except that this is part of my certificate "torture test". I have discovered a case where OpenSSL+libp11 not only doesn't work, but *crashes* in a situation where the GnuTLS build of openconnect works fine. You're basically asking me to remove that particular unit test from the suite because it fails. That isn't really how test suites are supposed to work :)

And this unit test wasn't even one of the ones where I was being deliberately nasty — I *expected* this one to pass. It happened on my *first* test of EC keys, when all I did was import them with p11tool in the manner which seemed natural. Now, I'm happy to go and prod @nmav and make sure that future versions of p11tool *will* import the pubkey as a separate object. But right now it's *not* an unlikely occurrence.

It would be better for users if the match *is* actually checked, and they get a coherent error message about a mismatch instead of a later SSL negotiation failure and an opaque 'TLS alert received' message. And if it's slightly slower, only in the case where the pubkey *isn't* present on the token, I think that's tolerable.

**Comment by dwmw2** [comment]
I've added the failing test case to OpenConnect in commit [e048030f8](http://git.infradead.org/users/dwmw2/openconnect.git/commitdiff/e048030f8). In fact the easiest way to reproduce is not to use OpenConnect at all but just to use its test harness.

• `git clone git://git.infradead.org/users/dwmw2/openconnect`
• `cd openconnect`
• `./autogen.sh && ./configure` — no need to make it; you only need `softhsm2.conf` to be created by  the configure script.
• `cd tests`
• `SOFTHSM2_CONF=softhsm2.conf gdb --args openssl s_client -engine pkcs11 -keyform engine -key 'pkcs11:token=openconnect-test1;object=EC;pin-value=1234' -cert certs/ec-cert.pem -connect localhost:443`

This assumes that SoftHSMv2 is installed correctly with a p11-kit module file. You can add '`HOME=.`' to the test run if not which will make p11-kit pick up the file from  `tests/.config/pkcs11/modules/softhsm2.module`. You might also need to edit that file to point to the right path for ibsofthsm2.so.

**Comment by nmav** [comment]
> You, the user, said you did not want to put the pubkey on the token.
> If you put the pubkey on the token, the problem goes away. 

Note that there are cards which have no public keys in them, only certificates.

**Comment by levitte** [comment]
Errr... The certificate contains a public key, anything else makes the cert useless...

**Comment by dwmw2** [comment]
Yes. The cert contains a public key. But in `X509_check_private_key()` you appear to be comparing the public key associated with the privkey, with the public key in the cert. Even when the former doesn't exist and you crash with a NULL pointer dereference...

**Comment by dengert** [comment]
@dwmw2  said: "No reason... except that this is part of my certificate "torture test". OK, then the tests have uncovered a problem, OpenSSL segfaults. it has also uncovered issues between different packages and assumptions they have made based on the standards they use. A user could fall into one of these issues.

**Comment by dwmw2** [comment]
@dengert In case you hadn't noticed, I'm trying to resolve *all* those issues, tilting at one windmill at a time. With a key imported into my token I expect to be able to use it from any application in my system by its RFC7512 ID. So all these _(and more)_ should JustWork™:
```
 openconnect -c 'pkcs11:token=foo;id=%01;pin-value=1234' vpn.example.com
 curl -E 'pkcs11:token=foo;id=%01;pin-value=1234' https://www.example.com
 nmcli con modify OfficeWifi 802-1x.private-key 'pkcs11:token=foo;id=%01;pin-value=1234'
```

**Comment by dengert** [comment]
@nmav said: "Note that there are cards which have no public keys in them, only certificates."
I agree.  But how should this  be handled? 

Should libp11 look for a matching certificate on the card if no public key can be found?

Should the PKCS#11 module look for the certificate to extrace the public key if no public is found on the card? Should it do this just to get the ec_point?

Should the card driver do the same?    

As I said the opensc PIV driver does. It needs it internally. It was also the card used to develop the OpenSC EC support, as well as the EC changes to libp11, and the opensc-engine and the mods to get OpenSSL to support an engine for ECDSA and ECDH. 

During that development, it was always assumed the card (or card driver) would present a public key  for each private key and libp11 could always get the ec-point from the public key to fill in the EVP_KEY. 

 Other drivers may or may not. That is upto the driver developer.

**Comment by levitte** [comment]
The way things currently are, openssl requires an engine to fill in the public key components in the EVP_PKEY before returning it. If the pkcs11 engine doesn't, it's currently faulty vis-a-vis openssl. Simple as that. 

You can certainly propose a change regarding this, but you must understand that this is a *big* semantic change, and there is no way you will see it before openssl 1.2 at the earliest. Among others, it will require an changed engine API. 

Side note: I agree that a pkcs11 client should be incorporated into openssl. I cannot predict in what version that would happen, that depends entirely on what API changes that would require...

**Comment by dengert** [comment]
@dwmw2 And you seam have the attention of representatives from most of the windmills.

But I have not seen anyone from softhsm which represents a card, card driver, and PKCS#11 module of its own.

**Comment by levitte** [comment]
Guys, name calling and blame game isn't very productive and really has no place here. Can we choose the path of moving forward instead?

**Comment by dengert** [comment]
@levitte  Even if a pkcs11 client is incorporated into openssl  ECC,  DSA and  GOST R 34.10-2001 PKCS#11 private keys do not contain the "DER-encoding of ANSI X9.62 ECPoint value Q", "Public value y" or  "64 bytes for public key; 32 bytes for each coordinates X and Y of elliptic curve point P(X, Y) in little endian order" respectively.  

Its only RSA PKCS#11 private keys that contain CKA_MODULUS and CKA_PUBLIC_EXPONENT.

**Comment by dengert** [comment]
@levitte 
Sorry, I am not trying to name call or place blame. I am trying to point out there are differences caused by interpretation of various standards and how this lead to this situation and how it could be resolved.

**Comment by levitte** [comment]
> @levitte  Even if a pkcs11 client is incorporated into openssl  ECC,  DSA and  GOST R 34.10-2001 PKCS#11 private keys do not contain the "DER-encoding of ANSI X9.62 ECPoint value Q", "Public value y" or  "64 bytes for public key; 32 bytes for each coordinates X and Y of elliptic curve point P(X, Y) in little endian order" respectively.

Yes, understood.  However, is that data possible to retrieve given the same token?  In that case, that's what an openssl engine is currently required to do when returning the EVP_PKEY.  If that information isn't possible to retrieve, then we do have a much more serious problem for sure.

**Comment by dengert** [comment]
The data may not be retrieval from the token. That depends on the token, how it was initialized and the how far the PKCS#11 module will (or can) go to get the information from the public key or certificate if they are on the token.  There are cases where data from the card may be cached or stored externally as in the PIV retired key case. NIST defined a Key History Object that can contain a offCardCertURL where these certs can be found.  The OpenSC PIV driver will not currently download the URL, but if the user has downloaded it to ~/.eid/cache/<sha1 name from url> the driver will read the certs as if they were on the card. So it is very token specific. 

The engine and libp11 usually only get data from the PKCS#11 module. But the certificate or public key may be available to OpenSSL via some other means. i.e. User has stored the certificate in a file that is read via application that calls OpenSSL. I believe this is what @dwmw2  is doing in his torture test, by including the parameter --servercert sha1:a82547f68f44d6351bef6cacd1d7b96e84f9dfa3

And that cert and the private key from libp11 are being passed to X509_check_private_key.
In most cases the the user considers these to be the matching pair. But a "torture test" should also include a test were they don't match.

**Comment by dwmw2** [comment]
With regard to native PKCS#11 support in OpenSSL, I think we already have a vague consensus on the path forward. We've talked in the not-so-distant past about merging something that looks a bit like the existing libp11, "after 1.1". Which I suppose means I need to pull my finger out and start looking at specifics some time soon :)
I have already obtained permission from the major contributors to libp11, for reusing their work under the OpenSSL licence.

That's kind of orthogonal to this specific issue... to which I think the answer *has* to be that when the private key is in hardware, you can't rely on having any additional information about it. When you're asked to use a given private key with `SSL_CTX_use_PrivateKey()`, you can reasonably assume that you can perform a 'sign' operation with it. But other than that, all bets are off. If you want to validate that it matches the X.509 certificate you were also asked to use for that same `SSL_CTX`, then sign some dummy data and validate that signature.

As for fixing this in 1.1 and earlier, vs. declaring it a more intrusive semantics change that has to wait for 1.2... if it is such a hard and fast rule that private keys must *always* have public keys attached, why does `EVP_PKEY_missing_parameters()` even exist? Could we not at least prevent `X509_check_private_key()` from segfaulting *somehow* in the older releases, to allow this use case to work? Changing semantics to always require a pubkey when previously we didn't, would be a change we can't make. But making it *more* tolerant shouldn't be so bad, surely?

Note, again, that this was not intended to be the *"torture"* part of my torture test — this was my first simple attempt at importing an EC key into SoftHSM with `p11tool` and trying to use it. I didn't *deliberately* omit the public key; that's just how it happened by default.

**Comment by dwmw2** [comment]
@dengert: No, the `--servercert sha1:…` is not relevant here; that's just telling the SSL client to accept the *server's* certificate with that fingerprint. Given that the server in our test suite is using a cert signed by our own test CA, I could have provided a `--cafile` argument. Or even just `--no-cert-check` although I'm seriously tempted to remove that option completely... people keep *using* it :)

Back on-topic... we use the `-c` argument to openconnect to set the client cert, and also perhaps the `-k` argument to specify the private key, if it isn't found in the same file or by the same (RFC7512) PKCS#11 URI as the cert. There is no option to provide a public key separately from the cert + privkey.

**Comment by dengert** [comment]
OK.  But the EC_POINT can be obtained from the pubkey that is in the cert.  But in what cases would it do this? Only if the private key was from an engine? Or flag from an engine that it was OK to do this?

**Comment by nmav** [comment]
> Note, again, that this was not intended to be the "torture" part of my torture test — this was my first simple attempt at importing an EC key into SoftHSM with p11tool and trying to use it. I didn't deliberately omit the public key; that's just how it happened by default.

May not be that accurate. I bet that the public key was there, but it was marked private (i.e., needed PIN to see it) by default. https://github.com/opendnssec/SoftHSMv2/pull/222

**Comment by dwmw2** [comment]
No I really this it was absent. I had already added --no-mark-private to work around https://github.com/OpenSC/libp11/issues/99 

Importing with sofrhsm2-util results in a public key object. Importing with p11tool seems not to. The openconnect suite has all three variants now. Two of which fail for the OpenSSL/libp11 build.

**Comment by levitte** [comment]
Hmmm...  a quick, cheap and dirty way would be for the pcks11 engine to copy the EVP_PKEY_METHOD of a newly allocated EVP_PKEY of the right type, then set it's own comparison functions with the help of `EVP_PKEY_asn1_set_param` and/or `EVP_PKEY_asn1_set_public`, and possibly replace other method functions that would otherwise require the public key components to be present in the EVP_PKEY structure.  I haven't tested this myself, but since the EVP_PKEY_METHOD type exists, it can as well be used for this type of thing.

**Comment by dwmw2** [comment]
Hm, cute idea but...
```
int X509_check_private_key(X509 *x, EVP_PKEY *k)
{
    EVP_PKEY *xk;
    int ret;

    xk = X509_get_pubkey(x);

    if (xk)
        ret = EVP_PKEY_cmp(xk, k);
    ...
```
```
int EVP_PKEY_cmp(const EVP_PKEY *a, const EVP_PKEY *b)
{
    ...
    if (a->ameth) {
    ...
        if (a->ameth->pub_cmp)
            return a->ameth->pub_cmp(a, b);
```
... we use the `EVP_PKEY_METHOD` of the key we found in the X509, not the engine key we're comparing with.

In the meantime, we could at least make it not *crash*, by being a little more defensive:
```
--- a/crypto/ec/ec_ameth.c
+++ b/crypto/ec/ec_ameth.c
@@ -224,6 +224,8 @@ static int eckey_pub_cmp(const EVP_PKEY *a, const EVP_PKEY *b)
     const EC_GROUP *group = EC_KEY_get0_group(b->pkey.ec);
     const EC_POINT *pa = EC_KEY_get0_public_key(a->pkey.ec),
         *pb = EC_KEY_get0_public_key(b->pkey.ec);
+    if (!group || !pa || !pb)
+        return -2;
     r = EC_POINT_cmp(group, pa, pb, NULL);
     if (r == 0)
         return 1;
@@ -394,6 +396,8 @@ static int ec_cmp_parameters(const EVP_PKEY *a, const EVP_PKEY *b)
 {
     const EC_GROUP *group_a = EC_KEY_get0_group(a->pkey.ec),
         *group_b = EC_KEY_get0_group(b->pkey.ec);
+    if (!group_a || !group_b)
+        return -2;
     if (EC_GROUP_cmp(group_a, group_b, NULL))
         return 0;
     else
```
That won't fix it though; that'll just make it fail more gracefully than a segfault. We'll end up in the `-2` case in `X509_check_private_key()`, setting the `X509_R_UNKNOWN_KEY_TYPE` error.

Now *maybe* the answer is for that `-2` case in `X509_check_private_key()` to *try* the key, by signing some dummy data and then validating it against the X509.

However, I see this is actually handled differently for RSA keys which have the same problem. In that case we never end up in `X509_check_private_key()` at all, because we have this check in `ssl_set_pkey()`:
```
#ifndef OPENSSL_NO_RSA
        /*
         * Don't check the public/private key, this is mostly for smart
         * cards.
         */
        if ((pkey->type == EVP_PKEY_RSA) &&
            (RSA_flags(pkey->pkey.rsa) & RSA_METHOD_FLAG_NO_CHECK)) ;
        else
#endif
        if (!X509_check_private_key(c->pkeys[i].x509, pkey)) {
```
So maybe the cleanest and simplest fix here is to extend that `RSA_METHOD_FLAG_NO_CHECK` case to be something we can set generically for any `EVP_PKEY`.

**Comment by dwmw2** [comment]
> OK. But the EC_POINT can be obtained from the pubkey that is in the cert. But in what cases would it do this? Only if the private key was from an engine? Or flag from an engine that it was OK to do this?

@dengert No, at this point we don't have "**the**" certificate. There exist hardware devices which *only* do keys, and don't have certificate storage at all. I believe the OpenPGP cards work like this, because they're not really designed with X.509 in mind. And the OpenSSL TPM engine certainly does.

Using a key from an engine and a certificate from elsewhere is a perfectly valid use case. In `X509_check_private_key()` we have "**a**" certificate and a key, and we are checking that they are indeed a match. We cannot simply copy the information from one to the other and then say *"oh look, they do match!"*. That's kind of missing the point :)

We *can* perform a dummy signature using the key, and validate it using the cert. Or, as noted above with the `RSA_METHOD_FLAG_NO_CHECK` flags, we can simply bypass the check altogether.

**Comment by dengert** [comment]
The issue appears to come down to associating a private key on a token with a certificate that may not be stored on the token. It may be in a certificate store or provided by the user or application. With RSA it was easy, use the modulus. But PKCS#11 does not define a private key with a EC_POINT, it says get it from the public key (or public key in the certificate.)  It does suggest that matching cert, public key and private key all use the same CKA_ID so they can be matched up.  But when the certificate to be used is not obtained via PKCS#11 and the private key does not have the EC_POINT there is no way to match them up without additional matching information. There is no unique value between the certificate and the private key.  

The additional information may be know to the user who could provide a single private key and single certificate which the user knows match.

A single sign and multiple verify operations could be used to match a private key and certificate, but on slow tokens, and the need for the PIN  in my option this is not a viable option. (It could also be the token has said the EC key can only be used for ECDH and not ECDSA.) 

Libp11 is the bridge that used the CKA_ID to match a key to a certificate or public key in order to get the EC_POINT to present as part of the private key to OpenSSL or other software so they can use the EC_POINT to match keys and certificates. 

So in other words OpenSSL private keys can not have a NULL EC_POINT.  If the private key is coming from PKCS#11 the certificate or public key must come from libp11.  If they are not on the token, then libp11 needs an alternate source of the certificate or public key and be provided with the CKA_ID or  to be used to match to a private key on the token. 

Maybe rfc7512  and p11-kit could provide a way to get the certificate from a file, and the private key form the token so libp11 then gets the EC_POINT from the certificate and adds it to the private EVP_KEY
used withing OpenSSL.

**Comment by dwmw2** [comment]
> Maybe rfc7512 and p11-kit could provide a way to get the certificate from a file, and the private key form the token so libp11 then gets the EC_POINT from the certificate and adds it to the private EVP_KEY used withing OpenSSL. 

The PKCS#11 URI is just a standard form for an identifier string *(really a search term)* for locating PKCS#11 objects by various criteria including `CKA_ID`, `CKA_LABEL` and fields from `CK_TOKEN_INFO`. Something like this doesn't really live there — not even if we'd suggested it during the years when it was an Internet-Draft, and before its final publication as RFC7512.

As for exposing the certificate to libp11... I'd prefer not to. Applications already expect that certificates and keys can come from different locations, and provide config file or command line options to specify them separately. The key may come from libp11 or an engine, while the certificate comes from somewhere else entirely. The existing code model is that the certificate and key for SSL client authentication *only* come together when you install them into the `SSL_CTX` with `SSL_CTX_use_certificate()` and `SSL_CTX_use_PrivateKey()`. (Or into the `SSL` if you must, but the principle is the same).

So I do not like the idea of somehow providing the certificate to libp11 *"in advance"*, just for it to copy over the EC_POINT data that OpenSSL doesn't *actually* need; it only uses it to check if the certificate and the key are the same. 

It would be much better just to fix OpenSSL *not* to require the EC_POINT in the EVP_PKEY object. After all, it already copes with *precisely* the equivalent situation for RSA keys in an engine, via the `RSA_METHOD_FLAG_NO_CHECK` flag. That's all we need.

If we really can't do that in existing released versions of OpenSSL, then let's work around it in the application — which *does* already have the certificate. It sucks, but then applications already have to have *hundreds* of lines of code even just to coherently load certificates from various types of files, let alone PKCS#11. I was already on the verge of BSD-licensing the OpenConnect code to do so and publishing it as a library alongside my certificate torture tests — the addition of this would just be a drop in the ocean :)

It would go something like:
 •  Sign dummy data with the `EVP_PKEY`.
 • Validate signature against the `X509` and bail out if it matches
 • `o2i_ECPublicKey(EVP_PKEY_get1_EC_KEY(pkey), X509_get_pubkey(x509))`

Of course, OpenSSL could do that too if it wanted, when it found no EC_POINT in `X509_check_private_key()`.

**Comment by dengert** [comment]
I thing you may be on to something here.

 My argument was assuming that OpenSSL or some application might try and find a certificate from a set of certificates by expecting the EC_POINT to match.  I don't see any code in in OpenSSL that is doing this , but I may have missed it.

The fact that for RSA they have the RSA_METHOD_FLAG_NO_CHECK used only in ssl_set_pkey and comment says it is there "mostly for smart cards" and implies this problem has been seen before for RSA keys. It was in 0.9.7. 

Here is another flag that looks promising: EC_PKEY_NO_PUBKEY  
[https://www.openssl.org/docs/man1.0.2/crypto/d2i_ECPrivate_key.html ]
but when reading a ECPrivateKey, they generate the EC_POINT from the private key, I don't think that would work with any tokens.

**Comment by dwmw2** [comment]
I have worked around this in OpenConnect thus:
http://git.infradead.org/users/dwmw2/openconnect.git/commitdiff/85ff68a9722b

It's just another hundred-odd lines of code where I already have over a thousand lines in my `SSL_CTX_ffs_please_OpenSSL_just_use_these_certs_the_user_asked_me_to_use()` functions... it's almost in the noise. But I *will* be revisiting this whole thing with a pointy stick :)

**Comment by dwmw2** [comment]
Yes. The crash is fixed, and the remaining issue is dealt with in #1551
